src/signatures/secp256k1_ecdsa.rs

```diff
@@ -0,0 +1,210 @@
+        let mut rng = rand::rngs::OsRng;
+        Self(k256::ecdsa::SigningKey::random(&mut rng))
+    }
+
+    #[cfg(feature = "rand")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "rand")))]
+    pub fn generate_with<R: rand::CryptoRng + rand::RngCore>(rng: &mut R) -> Self {
+        Self(k256::ecdsa::SigningKey::random(rng))
+    }
+
+    pub fn public_key(&self) -> PublicKey {
+        PublicKey(k256::ecdsa::VerifyingKey::from(&self.0))
+    }
+
+    pub fn to_bytes(&self) -> [u8; SecretKey::LENGTH] {
+        self.0.to_bytes().into()
+    }
+
+    pub fn try_from_bytes(bytes: [u8; SecretKey::LENGTH]) -> crate::Result<Self> {
+        k256::ecdsa::SigningKey::from_bytes(&bytes.into())
+            .map_err(|_| crate::Error::ConvertError {
+                from: "bytes",
+                to: "secp256k1 ecdsa secret key",
+            })
+            .map(Self)
+    }
+
+    // Signature operation can fail for some reason.
+    // Not sure what to do. Panic? Retry?
+    pub fn try_sign(&self, msg: &[u8]) -> crate::Result<Signature> {
+        self.0
+            .sign_recoverable(msg)
+            .map_err(|_| crate::Error::SignatureError { alg: "secp256k1 ecdsa" })
+            .map(|(sig, rid)| Signature(sig, rid))
+    }
+
+    pub fn sign(&self, msg: &[u8]) -> Signature {
+        self.try_sign(msg).expect("secp256k1 ecdsa sign failed")
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct PublicKey(k256::ecdsa::VerifyingKey);
+
+impl PublicKey {
+    pub const LENGTH: usize = 33;
+
+    pub fn verify(&self, sig: &Signature, msg: &[u8]) -> bool {
+        use k256::ecdsa::signature::Verifier;
+        self.0.verify(msg, &sig.0).is_ok()
+    }
+
+    pub fn to_bytes(self) -> [u8; PublicKey::LENGTH] {
+        let encoded_point = self.0.to_encoded_point(true);
+        let slice = encoded_point.as_bytes();
+        let mut bytes = [0_u8; PublicKey::LENGTH];
+        bytes.copy_from_slice(slice);
+        bytes
+    }
+
+    pub fn try_from_bytes(bytes: &[u8; PublicKey::LENGTH]) -> crate::Result<Self> {
+        k256::ecdsa::VerifyingKey::from_sec1_bytes(bytes)
+            .map(Self)
+            .map_err(|_| crate::Error::ConvertError {
+                from: "compressed bytes",
+                to: "Secp256k1 public key",
+            })
+    }
+
+    pub fn try_from_slice(bytes: &[u8]) -> crate::Result<Self> {
+        k256::ecdsa::VerifyingKey::from_sec1_bytes(bytes)
+            .map(Self)
+            .map_err(|_| crate::Error::ConvertError {
+                from: "compressed slice",
+                to: "Secp256k1 public key",
+            })
+    }
+
+    // credit: [secret_key_to_address](https://github.com/gakonst/ethers-rs/)
+    pub fn to_address(&self) -> Address {
+        // let public_key = secret_key.verifying_key();
+        let public_key = self.0.to_encoded_point(/* compress = */ false);
+        let public_key = public_key.as_bytes();
+        debug_assert_eq!(public_key[0], 0x04);
```

Comment:
For my curiosity, what does this type byte describe?

Reply:
As I mentioned above, SEC1 encoded point has the first byte as tag. Tag=4 encodes uncompressed point (public key).